### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by SaifAbuR3d**
This change is [Reviewable](https://reviewable.io/reviews/SaifAbuR3d/IMS/3)

### CODE REVIEW

File: IMS.BL/IMS.BL.csproj [hunk 1]
```diff
@@ -7,4 +7,8 @@
     <Nullable>enable</Nullable>
   </PropertyGroup>
 
+  <ItemGroup>
+    <ProjectReference Include="..\Utility\Utility.csproj" />
+  </ItemGroup>
+
```
Comment (ibrahim-masri) [inline IMS.BL/IMS.BL.csproj]: what does this do?

File: IMS.BL/Inventory.cs [hunk 2]
```diff
@@ -1,17 +1,106 @@
...
         public Inventory()
         {
             Products = new List<Product>();
         }
-        public List<Product> Products { get; set; }
+        private List<Product> Products { get; }
+
+        public bool AddProduct(Product product)
+        {
+            if (product == null)
```
Comment (ibrahim-masri) [inline IMS.BL/Inventory.cs]: How can I gurantee this not to be null in compile time

File: IMS.BL/Inventory.cs [hunk 3]
```diff
@@ -1,17 +1,106 @@
...
+            {
+                throw new ArgumentNullException("product cannot be null.");
+            }
+            if (Products.Exists(_product => _product.Name == product.Name))
+            {
+                throw new InvalidOperationException("The product name already exists.");
+            }
+
+            Products.Add(product);
+            Console.WriteLine("The product has been added successfully");
```
Comment (ibrahim-masri) [inline IMS.BL/Inventory.cs]: is this part of this class responsibility?

File: IMS.BL/Inventory.cs [hunk 4]
```diff
@@ -1,17 +1,106 @@
...
+            return true;
+        }
+        public bool AddProduct(string name, decimal price, int quantity)
```
Comment (ibrahim-masri) [inline IMS.BL/Inventory.cs]: why not one method?

File: IMS.BL/Inventory.cs [hunk 5]
```diff
@@ -1,17 +1,106 @@
...
+            }
+            foreach (var product in Products)
+            {
+                Console.WriteLine(product);
+            }
+        }
+
+        private Product? FindProduct(string productName)
+        {
+            var product = Products.Find(_product => _product.Name == productName);
```
Comment (ibrahim-masri) [inline IMS.BL/Inventory.cs]: what's the difference betweem .Find and .FirstOrDefault

File: IMS.BL/Inventory.cs [hunk 6]
```diff
@@ -1,17 +1,106 @@
...
+            {
+                Console.WriteLine(product);
+            }
+        }
+
+        public void UpdateProduct(string productName, string? newProductName,
+                                  decimal? newProductPrice, int? newProductQuantity)
+        {
+            var product = FindProduct(productName);
+            if (product == null)
```
Comment (ibrahim-masri) [inline IMS.BL/Inventory.cs]: why do use pattern matching sometimes and other times === null?

File: Utility/UserInputHandler.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+    public static decimal GetValidDecimal(string output)
+    {
+        decimal? input = null;
+        while (!input.HasValue)
+        {
+            Console.Write(output);
+            input = decimal.TryParse(Console.ReadLine(), out decimal d) ? d : new decimal?();
+        }
+        return (decimal)input;
```
Comment (ibrahim-masri) [inline Utility/UserInputHandler.cs]: is this casting needed?

File: Utility/UserInputHandler.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
+﻿// can add (optional) error messages and further input (range) validation
+public static class UserInputHandler
+{
+
+    public static decimal GetValidDecimal(string output)
+    {
+        decimal? input = null;
+        while (!input.HasValue)
+        {
+            Console.Write(output);
```
Comment (ibrahim-masri) [inline Utility/UserInputHandler.cs]: why are we printing this out?

File: Utility/UserInputHandler.cs [hunk 9]
```diff
@@ -0,0 +1,37 @@
+﻿// can add (optional) error messages and further input (range) validation
+public static class UserInputHandler
+{
+
+    public static decimal GetValidDecimal(string output)
+    {
+        decimal? input = null;
+        while (!input.HasValue)
```
Comment (ibrahim-masri) [inline Utility/UserInputHandler.cs]: wouldn't this be confusing to the user?

File: IMS.BL/Program.cs [hunk 10]
```diff
@@ -1 +1,70 @@
-﻿
\ No newline at end of file
+﻿using IMS.BL;
+
+namespace IMS.UI
+{
+    public class Program
+    {
+        static void Main()
```
Comment (ibrahim-masri) [inline IMS.BL/Program.cs:10]: same here, what is the responsibility for main?

File: IMS.BL/Program.cs [hunk 11]
```diff
@@ -1 +1,70 @@
...
+                        string productToSearch = UserInputHandler.GetValidString("Enter product name to search: ");
+                        inventory.SearchProduct(productToSearch);
+                        break;
+                    case 6:
+                        exit = true;
+                        break;
+                    default:
+                        Console.WriteLine("Invalid choice. Please try again.");
+                        break;
+                }
```
Comment (ibrahim-masri) [inline IMS.BL/Program.cs]: how can we have this in a more functional way?

File: IMS.BL/Inventory.cs (hunk 4 above)
Comment (SaifAbuR3d) [reply IMS.BL/Inventory.cs]: I think adding more variations (overloads) to a method can provide a more flexible interface to the class

File: Utility/UserInputHandler.cs (hunk 9 above)
Comment (SaifAbuR3d) [reply Utility/UserInputHandler.cs]: a new version of the Class UserInputHelper.cs with more meaningful names and clear intent: https://github.com/SaifAbuR3d/ATB/blob/master/ATB.BL/Helpers/UserInputHelper.cs

File: Utility/UserInputHandler.cs (hunk 7 above)
Comment (SaifAbuR3d) [reply Utility/UserInputHandler.cs]: casting decimal? to decimal, otherwise, it gives a compilation error

**Pull request merged**